File: src-ts/leetcode/gas-station.ts
function canCompleteCircuit(gas: number[], cost: number[]): number {
  let gasTank = 0;
  let total = 0;
  let start = 0;

  for (let stationIdx = 0; stationIdx < gas.length; stationIdx++) {
    const consume = gas[stationIdx] - cost[stationIdx];
    gasTank += consume;
    total += consume;
    if (gasTank < 0) {
      start = stationIdx + 1;
      gasTank = 0;
    }
  }

  return total < 0 ? -1 : start;
}

const add = (a: number, b: number) => a + b;
function canCompleteCircuitAlter(gas: number[], cost: number[]): number {
  const gasSum = gas.reduce(add);
  const costSum = cost.reduce(add);

  if (gasSum < costSum) return -1;

  let start = 0;
  let gasTank = 0;

  for (let stationIdx = 0; stationIdx < gas.length; stationIdx++) {
    gasTank += gas[stationIdx] - cost[stationIdx];

    if (gasTank < 0) {
      start = stationIdx + 1;
      gasTank = 0;
    }
  }

  return start;
}

function canCompleteCircuitBF(gas: number[], cost: number[]): number {
  const n = gas.length;

  for (let start = 0; start < n; start++) {
    let gasTank = 0;
    let station = start;
    for (station = start; station < n + start; station++) {
      gasTank += gas[station % n] - cost[station % n];

      if (gasTank < 0) break;
    }

    if (station === n + start && gasTank >= 0) return start;
  }

  return -1;
}

File: src-cpp/leetcode/gas-station.cpp
#include <bits/stdc++.h>
using namespace std;

int canCompleteCircuit(vector<int> &gas, vector<int> &cost)
{
    int gasTank = 0, total = 0, start = 0;

    for (int stationIdx = 0; stationIdx < gas.size(); stationIdx++)
    {
        int consume = gas[stationIdx] - cost[stationIdx];
        gasTank += consume;
        total += consume;
        if (gasTank < 0)
        {
            start = stationIdx + 1;
            gasTank = 0;
        }
    }

    return total < 0 ? -1 : start;
}

int canCompleteCircuitAlter(vector<int> &gas, vector<int> &cost)
{
    auto gasSum = accumulate(begin(gas), end(gas), 0);
    auto costSum = accumulate(begin(cost), end(cost), 0);

    if (gasSum < costSum)
        return -1;

    int gasTank = 0, start = 0;

    for (int stationIdx = 0; stationIdx < gas.size(); stationIdx++)
    {
        gasTank += gas[stationIdx] - cost[stationIdx];

        if (gasTank < 0)
        {
            start = stationIdx + 1;
            gasTank = 0;
        }
    }

    return start;
}

int canCompleteCircuitAlter2(vector<int> &gas, vector<int> &cost)
{
    int gasTank = 0, start = 0;

    vector<int> diff(gas.size());
    std::transform(begin(gas), end(gas), begin(cost), begin(diff), std::minus<int>());
    auto total = accumulate(begin(diff), end(diff), 0);
    if (total < 0)
        return -1;

    for (int stationIdx = 0; stationIdx < gas.size(); stationIdx++)
    {
        int consume = gas[stationIdx] - cost[stationIdx];
        gasTank += consume;

        if (gasTank < 0)
        {
            start = stationIdx + 1;
            gasTank = 0;
        }
    }

    return start;
}

// Brute Force
int canCompleteCircuitBF(vector<int> &gas, vector<int> &cost)
{
    int n = gas.size();

    for (int start = 0; start < n; start++)
    {
        int gasTank = 0;
        int station = start;
        for (station = start; station < n + start; station++)
        {
            gasTank += gas[station % n] - cost[station % n];

            if (gasTank < 0)
                break;
        }

        if (station == n + start && gasTank >= 0)
            return start;
    }

    return -1;
}